### ISSUE TITLE
JWT plugin is failing to verify a valid signature

### ISSUE THREAD

**Issue by kdstew** [Mashape/kong-dist-cloudformation#5]
On a kong 0.5.2 instance built using this cloudformation setup I've walked through the setup of the JWT plugin [here](https://getkong.org/plugins/jwt/) but when I create a jwt token and try to make a request I get back `{"message":"Invalid signature"}`.  

```
secret: ad30a8c4625a446bc44d2d1c764be04a
key: 94e96aa0ab934c2dcc9c5ff910a29748

curl <host>:8000/rise -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiI5NGU5NmFhMGFiOTM0YzJkY2M5YzVmZjkxMGEyOTc0OCJ9.qkdJ0FYTZiIQGHe1w8g1Ta-Y1ilWoX2BfHzkvFJGQY0'
```

To troubleshoot I setup kong 0.5.2 and cassandra locally using docker as described [here](https://github.com/Mashape/docker-kong).  Then I went through the same steps as I did with the ec2 setup.  When I curled the endpoint with the token it was verified and I received the expected response.

To try to narrow down what was causing the issue I went through the [steps of setting up the kong development environment](https://github.com/Mashape/kong#development) on the ec2 instance created using cloudformation.  Then I added some `ngx.log` statements in `kong/plugins/jwt/jwt_parser.lua`.  

```
local alg_verify = {
  ["HS256"] = function(data, signature, key)
    ngx.log(ngx.ERR, "Signature: "..signature);
    ngx.log(ngx.ERR, "alg_sign['HS256']: "..alg_sign["HS256"](data, key));
    ngx.log(ngx.ERR, "call crypto direct: "..crypto.hmac.digest("sha256", data, key, true) );
    ngx.log(ngx.ERR, "call crypto direct(hex): "..crypto.hmac.digest("sha256", data, key, false) );
    ngx.log(ngx.ERR, "calling with actual values(hex): "..crypto.hmac.digest("sha256", 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiI5NGU5NmFhMGFiOTM0YzJkY2M5YzVmZjkxMGEyOTc0OCJ9', 'ad30a8c4625a446bc44d2d1c764be04a', false) );

    ngx.log(ngx.ERR, "data, key: "..data..", "..key);
    ngx.log(ngx.ERR, signature == alg_sign["HS256"](data, key));
    return signature == alg_sign["HS256"](data, key)
  end
  --["HS384"] = function(data, signature, key) return signature == alg_sign["HS384"](data, key) end,
  --["HS512"] = function(data, signature, key) return signature == alg_sign["HS512"](data, key) end,
}
```

This showed me that the value returned by `crypto.hmac.digest("sha256", data, key, true)` was not what it should be, it should match the value of Signature.  The follow was logged.

```
Signature: �GI�Vf"w���5M���)V�}�||�RFA�
alg_sign['HS256']: 
call crypto direct: 
call crypto direct(hex): 0000000000000000000000000000000000000000000000000000000000000000
calling with actual values(hex): 0000000000000000000000000000000000000000000000000000000000000000
data, key: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiI5NGU5NmFhMGFiOTM0YzJkY2M5YzVmZjkxMGEyOTc0OCJ9, ad30a8c4625a446bc44d2d1c764be04a
false

```

The value calculated by crypto should match `signature` but instead shows as nothing, by changing the last param in the crypto call to `false` to return the value as hex, I got `0000000000000000000000000000000000000000000000000000000000000000`.  But if you convert the value of signature to hex I got this `aa4749d056136622101877b5c3c8354daf98d62956a17d817c7ce4bc5246418d`.

If I run that same crypto command `crypto.hmac.digest("sha256", data, key, false)` with the same data and key values locally I get the expected hex result of `aa4749d056136622101877b5c3c8354daf98d62956a17d817c7ce4bc5246418d`.

I compared the luacrypto library versions on the ec2 instance and locally and they are the same.  Looking at the [luacrypto code](https://github.com/mkottman/luacrypto) it looks like it uses the openssl libraries installed on the server.  This is where the two servers differed slightly.

```
ec2 instance: 
openssl version
    OpenSSL 1.0.1k-fips 8 Jan 2015
sudo yum list installed | grep openssl
    openssl.x86_64                        1:1.0.1k-10.87.amzn1         @amzn-main
    openssl098e.x86_64                    0.9.8e-29.18.amzn1           @amzn-main

docker instance:
openssl version
    OpenSSL 1.0.1e-fips 11 Feb 2013
sudo yum list installed | grep openssl
    openssl.x86_64                  1:1.0.1e-42.el7.9        @updates
    openssl-libs.x86_64             1:1.0.1e-42.el7.9        @Updates
    openssl098e.x86_64              0.9.8e-29.el7.centos.2   @base
```

I believe the issue I am experiencing is being caused by the version of openssl installed on the ec2 instance.

**Comment by subnetmarco** [comment]
On a side note, I noticed that luacrypto hasn't been maintained in a while, and we should probably consider using something more up to date like http://25thandclement.com/~william/projects/luaossl.html

Regarding the issue, just to understand better, basically one version of OpenSSL returns the right value, and the other one just returns `0000000000000000000000000000000000000000000000000000000000000000`?

Which OpenSSL version doesn't seem to be working?

**Comment by kdstew** [comment]
Correct.  As far as I could tell the openSSL version was the one piece that was different on the machines.  The version of openSSL on the server where I'm getting the `0000000000000000000000000000000000000000000000000000000000000000` result is `OpenSSL 1.0.1k-fips 8 Jan 2015` and that is on the ec2 box created by cloudformation.

**Comment by subnetmarco** [comment]
Just out of curiosity, can you try executing openssl manually on both machines and see what happens:

``` bash
echo -n "value" | openssl dgst -sha256 -hmac "key"
```

**Comment by kdstew** [comment]
on machine returning the bad result:

```
echo -n "value" | openssl dgst -sha256 -hmac "key"
(stdin)= 90fbfcf15e74a36b89dbdb2a721d9aecffdfdddc5c83e27f7592594f71932481
```

on the machine returning the expected result:

```
echo -n "value" | openssl dgst -sha256 -hmac "key"
(stdin)= 90fbfcf15e74a36b89dbdb2a721d9aecffdfdddc5c83e27f7592594f71932481
```

**Comment by kdstew** [comment]
So out of curiosity I wondered if reinstalling the luacrypto lib would fix the issue.  The following is what I did on the box where signature validation was failing.

```
sudo su -

luarocks remove luacrypto --force
  Checking stability of dependencies on the absence of
  luacrypto 0.3.2-1...

  The following packages may be broken by this forced removal:
  kong 0.5.2-1
  lapis 1.1.0-1

  Removing luacrypto 0.3.2-1...
  Missing dependency for lapis 1.1.0-1: luacrypto
  Missing dependency for kong 0.5.2-1: luacrypto >= 0.3.2-1
  Missing dependency for kong 0.5.2-1: luacrypto
  Removal successful.

luarocks install luacrypto
  # install failed because it could not find openssl/evp.h

yum install openssl-devel

luarocks install luacrypto
  # install failed because it could not find gcc

yum install gcc

luarocks install luacrypto

kong stop

kong start

curl localhost:8000/rise -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiI5NGU5NmFhMGFiOTM0YzJkY2M5YzVmZjkxMGEyOTc0OCJ9.qkdJ0FYTZiIQGHe1w8g1Ta-Y1ilWoX2BfHzkvFJGQY0'
  {"started":"2015-10-06T17:23:00.343Z","uptime":3813500.72}
```

Since the openssl-devel files were not on this box I'm not really sure what version of openssl luacrypto was built with.  But by adding the `openssl-devel 1:1.0.1k-10.87.amzn1` package and reinstalling/rebuilding luacrypto I am able to get the Kong JWT plugin working.

**Comment by subnetmarco** [comment]
Basically we are generating the AWS image by taking our CentOS 5 distribution (because AWS Linux AMI is originally a modified version of CentOS 5) - I guess we need to change the way we build the AWS distribution by running it directly on AWS, to avoid any problem.

**Comment by arsdehnel** [comment]
I'm running into this same issue with an HVM AMI that I just setup last week, is there something I need to do to that instance to fix this? It sounds to me like the AMI was fixed back in January according to the above comment but I'm having the same behavior where my local Docker setup is accepting my token as valid but the AWS instance is saying it's invalid.

```
secret: 97f81d1594494d5eb2421b2e114097b9
key: 0342c7e0280d4639849e0ba960fd1b46

curl 192.168.99.101:8000/tests/jwt -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiIwMzQyYzdlMDI4MGQ0NjM5ODQ5ZTBiYTk2MGZkMWI0NiIsImlhdCI6MTQ2MDI4NTY0NX0.GmAxo87RWs9Kl9KNqO9on06zfBTDdUeYqWRw2Lc7fx0'
```

**Comment by subnetmarco** [comment]
@arsdehnel if you are using the Amazon Linux operating system, can you try to manually install the AWS package on the instance from https://getkong.org/install/aws-linux/ ?

The problem was a mismatch for the OpenSSL library that was causing an exception, but the AWS distribution should have fixed this problem a long time ago.

**Comment by shashiranjan84** [comment]
@thefosk Can I close this ticket?

**Comment by cvrabie** [comment]
Just so you know, the AWS Marketplace image still has this problem. Using @kdstew 's solution does solve the problem.

**Comment by subnetmarco** [comment]
@cvrabie have you tried to manually install the AWS Linux AMI distribution: https://getkong.org/install/aws-linux/

**Comment by shashiranjan84** [comment]
@cvrabie which version of Kong AWS Marketplace image did you try?

**Comment by cvrabie** [comment]
@shashiranjan84 just selected the one from your website. it gave me this one: kong-0.8.1-cassandra-2.2.4-hvm-8e007d1a-73c7-4303-a5a8-ab26fa71e02e-ami-8431dce9.3 (ami-bc8903cf)

**Comment by cvrabie** [comment]
@thefosk i didn't try to remove it then reinstall it but if I run `yum update` it says everything is up to date

**Comment by subnetmarco** [comment]
I confirm that this issue is finally resolved in the latest 0.9 RC.

**Comment by shashiranjan84** [comment]
Closing this ticket as it has been fixed in 0.9.0.
